### ISSUE TITLE
머신러닝 기초 관련 (y의 수)

### ISSUE THREAD

**Issue by ji0eeeee** [codingeverybody/codingyahac#860]
질문 내용:
머신러닝을 할때 x의 값 경우 변수를 여러개로 두고 하나의 y값을 구할 수 있다고 알고있습니다.
혹시 x변수를 여러개 했을때 예를들면 y1,y2값을 구할수도 있나요?? 예를 들면 이런식의 데이터셋도 가능한가요?
![1111](https://user-images.githubusercontent.com/56102978/71707783-d039a200-2e2f-11ea-862b-ea00c7b4957c.PNG)

**Comment by RayleighKim** [comment]
안녕하세요.

1. **네, 그렇습니다.**
'예를 들면 이런식의 데이터셋도 가능한가요?'라는 문장을 보기 전까지는 그렇게 답할 수 있었습니다.
구조적으로 문제가 없습니다.
예를 들자면,
x변수 여러개(온도, 습도 ,풍속)를 가지고 y여러개(레몬에이드 판매량, 오렌지에이드 판매량)를
예측할 수는 있습니다.

2. 데이터셋 예시. **질문의 보완을 부탁드립니다.**
예시를 들어주신 데이터를 보았더니, 생각 하고 계시는 문제가 여러개가 아닌가하는 인상을 받았습니다.
ㄱ. 빈값도 많이 보이고, 
ㄴ. 특정 row만 Y3와 x2_data, x3_data, x4_data 가 있습니다.
질문을 조금 더 구체적으로 쪼개 주실 수 있을까요?
그렇게 해주신다면 더 자세히 답변을 다는데 도움이 됩니다.

**Comment by ji0eeeee** [comment]
안녕하세요. 답변 정말 감사드립니다.

- 해당 데이터 셋에서 빈 값은 작동을 안 한다는 뜻으로 생각하고 비워뒀습니다. 
- 저 같은 경우 X 변수를 시간, 감정, 날씨로 지정하였고 Y 변수를 색, 밝기, 색온도로 지정하였습니다.

![image](https://user-images.githubusercontent.com/56102978/71952179-225c3800-3221-11ea-9eff-c415ae256cac.png)

- 또 사진에 있는 엑셀 데이터를 불러오기 위해서 다음과 같은 코드를 작성하였는데 제대로 불러와지지 않는것 같아서 .. (해당 데이터의 경우 빈 값에 0.1을 넣고 불러온것 입니다)

![image](https://user-images.githubusercontent.com/56102978/71952243-57688a80-3221-11ea-9b71-d60a9b8dc9e7.png)
![image](https://user-images.githubusercontent.com/56102978/71952248-5cc5d500-3221-11ea-93aa-12a07632b506.png)

- 다음과 같은 코드를 사용해서 빈 값에는 0를 넣고 간단하게 돌려보았는데 제가 하는 방법이 맞는지 확신이 안 들어서요 ,, Y값이 한개일때 사용하는 코드를 보고 참고해서 작성한 것 입니다.

![image](https://user-images.githubusercontent.com/56102978/71952273-72d39580-3221-11ea-9d50-33ccec07e636.png)
- 결과가 다음과 같이 나왔는데 이렇게 나왔을때 X값이 2 2 2 라는 건가요? 그렇다면 Y값은 어떤거라는 건가요 .. ? 


- 배워가는 과정이라 많이 부족한데 친절한 답변 감사합니다 !!

**Comment by RayleighKim** [comment]
구체적인 코드를 보여주셔서 고맙습니다.
사실, 풀고 싶은 문제를 제가 자세히 알지 못하면 답변이 부웅 뜰 수 있는데
코드가 있다면 제가 조금 더 구체적으로 답변을 달 수 있어요.
문제 맥락과 상황을 알면 더욱 구체적으로 답할 수 있구요.

아래는 답변입니다.
-----------------------------------------
1. 빈 값은 정말 조심하셔야 합니다.
나의 의도와 달리, 빈값이 있으면 연산이 아예 안되기도 합니다.
빈 값에 0.1을 둔다면 그것도 수학적으로 의도가 명확해야 합니다.
빈 값이 있는 row를 아예 떼어야 할 수도 있고
빈 값이 있는 칸을 최대한 메워야할 수 도 있습니다.
전체, 빈칸있는row만, 빈칸없는row만 따로 모델링을 해야할 수도 있어요.

2. 저는, 엑셀데이터라면 pandas로 불러와서 처리를 다 한 뒤, .values를 통해 넘파이 어레이로 넘기는 걸 선호합니다.  엑셀 데이터가 제대로 안불러와지는 거 같다면 에러메세지를 바탕으로 검색해보고, 해결이 안되면 상황을 보여주시면 답변을 또 드릴께요.

3. 결과에 대하여.
Prediction : 뒤의 프린트 된 부분은 다음과 같으므로, 테스트 데이터에 대한 예측값을 출력하는 부분입니다.
```
sess.run(prediction,  feed_dict={X:x_test})
```
위에 prediction은 아래와 같이 정의하셨으므로, y가 1이 나오는 값의 위치를 보여줍니다.
```
prediction = tf.argmax(hypothesis, 1)
```
테스트 데이터는 전부 y가 2번째(0번째부터 세어서)에 있네요.

즉 아래처럼 출력된 것은, 테스트 데이터 3개에 대한 각각의 예측값입니다.
y가 1이되는 부분이 전부 2번째 라는 뜻이네요.
Prediction : [2 2 2]


4. logistic regression이고, 멀티클래스에 대해 일반화한 표현으로 softmaxregression입니다.
linear regression과 용도가 다릅니다.

5. 카테고리 데이터와 dummy,  멀티클래스label과 one-hot encoding등을 공부하기를 권장드려요.
변수가 시간, 감정, 날씨라면....
시간은 대부분 dummy를 쳐야 합니다. 순서에 의한 효과를 보고 싶어도 dummy를 치는 것이 더 옳은 경우가 많습니다. (물론, 모델특성에 따라 전략이 달라지기도 합니다.)
감정변수도 날씨변수도, 카테고리 데이터인지 수치적인 데이터인지에 따라
필요한 전처리, 전략이 바뀔거에요.

6. Y가 색, 밝기, 색온도
위 값들이 역시 수치데이터인지 카테고리데이터인지에 따라 이야기가 많이 달라집니다.
분류 문제가 아니라면, 지금 참고하신 코드는..
hypothesis 에서 tf.nn.softmax를 떼어야 맞을 거에요. cost도 수정해주셔야 합니다.
동시에, prediction, is_correct, accuracy 도 없애주셔야 합니다.

만일 분류 문제여도,  색에 1이 들어가도 밝기에 1이 들어갈 수 있다?
색,  밝기, 색온도에 대해서 모델을 각각 만들어주시는게 좋을 수도 있습니다.


고맙습니다.

**Comment by ji0eeeee** [comment]
안녕하세요. 독학으로 공부하는 중 어려움이 많았는데 친절한 답변 정말 감사합니다!

**Comment by RayleighKim** [comment]
Y테스트, Y데이터 모양을 보시면, n개의row에, 3개의 컬럼으로 이루어져있습니다.
각 row는 딱 한 컬럼에만1이 들어가고 나머지가 0인걸로 보아, 원핫인코딩된 y입니다.
0번째 자리, 1번째 자리, 2번째 자리각각 의미가 있습니다. 예를들면
1 0 0 : 개
0 1 0 : 고양이
0 0 1 : 원숭이
라든지 요.

위 코드의 argmax는, 가장 큰 값의 위치를 찾아줍니다.

그래서 실제 예측값은 0 0 1인데, 2번째 자리가 값이 가장 크므로, 2라고 예측값을 반환해줍니다.

테스트데이터의 예측값 세개가 전부 2 2 2 라고 나온 것은 위와 같은 맥락에서 해석해주세요.

**Comment by ji0eeeee** [comment]
답변 정말 감사합니다.
